Reuse already read lastLocation information during location updates

cshannon opened the pull request:
This change will reuse previously read lastLocation information when ManagerMetadataUtil updates last location to prevent unnecessary metadata reads. If the value is not known then a metadata read will be performed like before.

This closes #3301

cshannon commented:
@keith-turner - It seemed like the best thing to do here was to add a new field to the `Assignment` class to pass the previously read last location data if known. This was nice as I didn't have to change method signatures and also works for the case where a list of assignments are passed. I added a bunch of mock tests to verify that the new behavior is correct in terms of when the update will do a metadata read or skip it. I ran through the sunny tests and they all passed but I will also kick off a full IT build.

dlmarion commented on server/base/src/main/java/org/apache/accumulo/server/util/ManagerMetadataUtil.java:
```diff
@@ -257,14 +257,23 @@ public static Optional<StoredTabletFile> updateTabletDataFile(ServerContext cont
-      TabletMetadata lastMetadata = ample.readTablet(extent, TabletMetadata.ColumnType.LAST);
-      Location lastLocation = (lastMetadata == null ? null : lastMetadata.getLast());
+      // If prevLastLocation is not provided then look it up from metadata
+      final Location lastLocation = Optional.ofNullable(prevLastLocation).orElseGet(() -> {
```
Could use a conditional statement here and avoid the object creation.

```
  Location lastLocation = prevLastLocation;
  if (lastLocation == null) {
        log.trace(
            "Previous last location for {} not provided to updateLastForAssignmentMode, loading from metadata.", extent);
        final TabletMetadata lastMetadata =
            ample.readTablet(extent, TabletMetadata.ColumnType.LAST);
        lastLocation = lastMetadata == null ? null : lastMetadata.getLast();
  }
```

keith-turner commented on server/base/src/main/java/org/apache/accumulo/server/util/ManagerMetadataUtil.java:
```diff
@@ -257,14 +257,23 @@ public static Optional<StoredTabletFile> updateTabletDataFile(ServerContext cont
+        log.trace(
+            "Previous last location for {} not provided to updateLastForAssignmentMode, loading from metadata.",
+            extent);
+        final TabletMetadata lastMetadata =
```
If the last was not present when we read earlier, then we probably do not need to read it again.  Its tricky though because null can have two meanings, it can mean it was never read previously or it was read previously and was not present.  One simplification is that if we assume that all code calling this method has already tried to read the last location, then we never need to read it here when null.

ctubbsii commented:
The original ticket was targeting 2.1.1, but this PR was made against the main branch instead of the 2.1 branch. I'm not sure what the desired intent is, but these should probably match, or we'll need another PR to backport it if they remain different. If you want to avoid creating a new PR, this PR can be updated by editing the base branch to 2.1, and then rebasing (cherry-picking) the commits from the HEAD of the main branch onto the head of the 2.1 branch and force pushing the branch that backs this PR.

cshannon commented:
> The original ticket was targeting 2.1.1, but this PR was made against the main branch instead of the 2.1 branch. I'm not sure what the desired intent is, but these should probably match, or we'll need another PR to backport it if they remain different. If you want to avoid creating a new PR, this PR can be updated by editing the base branch to 2.1, and then rebasing (cherry-picking) the commits from the HEAD of the main branch onto the head of the 2.1 branch and force pushing the branch that backs this PR.

I actually didn't need to do anything here but switch the target merge branch in my PR to 2.1. I actually already did base this PR off the 2.1 branch (I thought I did) but when I created the PR I just mistakenly picked the wrong target as it defaulted to main and I didn't notice.

cshannon commented:
@keith-turner - This is ready for another review, I simplified things and removed the attempt to read the metadata if the passed in lastLocation value is null and updated the Assignment class to only have one constructor that requires lastLocation now so we don't miss passing in the value. All the sunny tests passed and I kicked off another full IT build as well.

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -924,7 +925,8 @@ private void getAssignmentsFromBalancer(TabletLists tLists,
 
-            tLists.assignments.add(new Assignment(assignment.getKey(), assignment.getValue()));
+            tLists.assignments.add(new Assignment(assignment.getKey(), assignment.getValue(),
+                lastLocation != null ? Location.last(lastLocation) : null));
```
Looking around in the code I am not entirely sure `lastLocation` is actually a location from the last column. Not sure what the value in the map is for ATM.  Entries are added to the unassigned map in the following method and that does not look like its adding things from the last column to the map.

https://github.com/apache/accumulo/blob/9c459ce8f06d9d9e04280d020ea091fedf13749b/server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java#L383-L399

cshannon replied:
I updated the PR to use the lastLocation value that is part of TabletLocationState when doing the metadata updates. Before the change the value was always coming from getServer() which is returning getLocation() and that could be current, future or last so this makes sure to always use the last value from metadata. The old value (getServer()) is still being tracked now as I created an UnassignedTablet class to track both that value and lastLocation as they are used for different things.

cshannon commented on server/base/src/main/java/org/apache/accumulo/server/util/ManagerMetadataUtil.java:
```diff
@@ -253,18 +253,16 @@ public static Optional<StoredTabletFile> updateTabletDataFile(ServerContext cont
+      Ample.TabletMutator tabletMutator, TServerInstance location, Location lastLocation) {
     // if the location mode is assignment, then preserve the current location in the last
     // location value
+    log.info("Loaded Last location: {}", lastLocation);
```
Yep, I forgot to take this out, I will remove it.

cshannon commented:
I ran a full IT build against this PR and it passed.

keith-turner commented on server/base/src/main/java/org/apache/accumulo/server/manager/state/UnassignedTablet.java:
```diff
@@ -0,0 +1,49 @@
+  public UnassignedTablet(Location location, Location lastLocation) {
+    Preconditions.checkArgument(
+        lastLocation == null || lastLocation.getType() == TabletMetadata.LocationType.LAST);
+    this.location = location;
```
If the following is also en expecation, could check it.

```suggestion
     Preconditions.checkArgument(location == null || location.getType() == TabletMetadata.LocationType.CURRENT || location.getType() == TabletMetadata.LocationType.FUTURE)
    this.location = location;
```

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -240,7 +242,7 @@ public void run() {
           });
           TabletGoalState goal = manager.getGoalState(tls, mergeStats.getMergeInfo());
-          TServerInstance location = tls.getServer();
+          Location location = tls.getLocation();
```
Later in the code the location is logged and the Location class does not seem to have a toString method, so with this change will need to add a toString method to the Location class or change the log stmt.

keith-turner commented on server/base/src/main/java/org/apache/accumulo/server/manager/state/UnassignedTablet.java:
```diff
@@ -0,0 +1,49 @@
+
+import com.google.common.base.Preconditions;
+
+public class UnassignedTablet {
```
I like this new class it makes the existing code more clear.  I didn't see anything logging this class, but not sure its not logged. If its logged anywhere it would need a toString.

cshannon replied:
I'll add toString() and also equals/hashcode so it's there for the future if we need it.
